database/CoreDatabase.java

```diff
@@ -540,51 +538,47 @@ private void validateIsolation(CoreTransaction.Data txn, CoreTransaction.Data ma
         }
 
         public void committed(CoreTransaction.Data txn) {
-            assert commitStates.get(txn) == CommitState.COMMITTING && txn.snapshotEnd().isPresent();
-            commitStates.put(txn, CommitState.COMMITTED);
-            commitTimeline.compute(txn.snapshotEnd().get(), (snapshot, committed) -> {
-                if (committed == null) committed = new HashSet<>();
-                committed.add(txn);
-                return committed;
-            });
+            assert committing.contains(txn) && txn.snapshotEnd().isPresent();
+            committed.add(txn);
+            committing.remove(txn);
         }
 
         void closed(CoreTransaction.Data txn) {
-            if (commitStates.get(txn) != CommitState.COMMITTED) commitStates.remove(txn);
+            // txn closed with commit or without failed commit
+            uncommitted.remove(txn);
+            committing.remove(txn);
             cleanupCommitted();
         }
 
         private void cleanupCommitted() {
             if (cleanupRunning.compareAndSet(false, true)) {
-                Optional<Long> oldestUncommittedSnapshot = oldestNotCommittedSnapshot();
-                ConcurrentNavigableMap<Long, Set<CoreTransaction.Data>> deletable;
-                if (oldestUncommittedSnapshot.isEmpty()) deletable = commitTimeline;
-                else deletable = commitTimeline.headMap(oldestUncommittedSnapshot.get());
-                deletable.keySet().forEach(snapshot ->
-                        deletable.computeIfPresent(snapshot, (s, txns) -> {
-                            for (CoreTransaction.Data txn : txns) {
-                                txn.delete();
-                                commitStates.remove(txn);
-                            }
-                            return null;
-                        })
-                );
+                long lastCommittedSnapshot = newestCommittedSnapshot();
+                Long cleanupUntil = oldestUncommittedSnapshot().orElse(lastCommittedSnapshot + 1);
+                committed.forEach(txn -> {
```

Comment:
cleanup is simplified by to only deleting the `committed` transactions up to a certain age